parlai/chat_service/core/chat_service_manager.py

```diff
@@ -0,0 +1,677 @@
+import parlai.chat_service.services.messenger.shared_utils as shared_utils
+from parlai.chat_service.services.messenger.world_runner import MessengerWorldRunner
+from abc import ABC, abstractmethod
+
+
+parent_dir = os.path.dirname(os.path.abspath(__file__))
+
+
+class AgentState:
+    """Keep track of Agent State.
+
+    State includes which is the "active" agent - i.e., which agent in which
+    world do we message, etc.
+    """
+
+    def __init__(self, service_id, overworld_agent):
+        self.service_id = service_id
+        self.overworld_agent = overworld_agent
+        self.active_agent = overworld_agent
+        self.task_id_to_agent = {}
+        self.onboard_data = None
+        self.stored_data = {}
+        self.time_in_pool = {}
+
+    def get_active_agent(self):
+        """Return active messenger agent.
+
+        :return:
+            a MessengerAgent, which corresponds to the active agent for this
+            agent state.
+        """
+        return self.active_agent
+
+    def set_active_agent(self, active_agent):
+        """Set active agent for this agent.
+
+        :param active_agent:
+            A MessengerAgent, the new active agent for this given agent state
+
+        """
+        self.active_agent = active_agent
+
+    def get_overworld_agent(self):
+        """Return overworld messenger agent.
+
+        :return:
+            a MessengerAgent, which corresponds agent object in the overworld
+        """
+        return self.overworld_agent
+
+    def get_id(self):
+        """Return agent's ID.
+
+        :return:
+            int agent ID
+        """
+        return self.service_id
+
+    def has_task(self, task_id):
+        """Determine if an agent is in a task.
+
+        :param task_id:
+            string task id
+
+        :return:
+            if agent is in that task.
+        """
+        return task_id in self.task_id_to_agent
+
+    def get_agent_for_task(self, task_id):
+        """Return MessengerAgent for given task id.
+
+        For each "player", a separate agent is created for each task. This
+        returns the appropriate MessengerAgent given the task id
+
+        :param task_id:
+            string, task id
+
+        :return:
+            messenger agent object associated with the given task
+        """
+        if self.has_task(task_id):
+            return self.task_id_to_agent[task_id]
+        else:
+            return None
+
+    def assign_agent_to_task(self, agent, task_id):
+        """Mark agent in task.
+
+        :param agent:
+            MessengerAgent object to mark in task
+        :param task_id:
+            string task name
+        """
+        self.task_id_to_agent[task_id] = agent
+
+
+class ChatServiceManager(ABC):
+    class ChatServiceMessageSender(ABC):
+        """ChatServiceMessageSender is a wrapper around requests that simplifies the
+        the process of sending content.
+        """
+
+        @abstractmethod
```

Comment:
are these the only two functions necessary for the sender? Also, does this imply that each subclass of `ChatServiceManager` must implement the `ChatServiceMessageSender`?

Reply:
Yep, these are the only two functions necessary –– all of the other functions were messenger specific. 

I think having this requirement is fine, the `MessengerSender` class has a lot of helper methods that would clutter the logic in `Manager` if they were implemented there instead –– I think there for any platform, there would be a lot of per-platform code that merits being in its own subclass.